### ISSUE TITLE
Question: How to use Arcade for building an emulator?

### ISSUE THREAD

**Issue by mmartinortiz** [pythonarcade/arcade#2439]
Hi,

Few months ago I embarked myself in the adventure of writing an emulator for the CHIP-8 system just for fun. At the moment of deciding what library to use for interacting with the host's keyboard and screen I decided to go for Arcade. Currently, I use the version `2.6.17`.

My emulator uses basically the following functions from an `arcade.Window` object:

- `__init__` for setting up the emulated machine: reset the memory, screen and registries.
- `setup` for loading a ROM for the CHIP-8.
- `on_draw` for drawing the screen.
- `on_key_xx` for registering keys pressed by the user.
- `update` to run a cycle of my emulated machine: fetch an opcode from the program counter, decode it, call the function that emulates the opcode operation and update the timers. That's it, there are no internal loops.

At the moment, I do not control the emulated machine speed, and it should run at the same speed than the host's computer. I thought to implement a mechanism for emulating the clock speed later in the process.

However, the emulator is too slow, the FPS reported by `arcade.get_fps()` is around 16. Here is an example of the current speed:

[Screencast from 2024-10-31 20-07-42.webm](https://github.com/user-attachments/assets/198871f4-471d-4d22-ae13-ee48a96505d6)

Using the `update_rate()` function does not have any effect as [described here](https://github.com/pythonarcade/arcade/issues/1336) so I make use of `pyglet.app.run()` directly, as described in the mentioned issue, but it does not matter what refresh rate I use, the FPS is always the same.

Am I using Arcade correctly for building an emulator? Do you think that I should consider to make use of Arcade in a different way?

**Comment by Cleptomania** [comment]
I'd suggest, if you can, upgrading to Arcade 3.0 pre-releases, they're more stable than 2.6 anyways, and have a lot of performance improvements.

The issue of the update rate and fps being tied together should be completely fixed in 3.0.

**Comment by pushfoo** [comment]
There seem to be some easy fixes you can make [here](https://github.com/mmartinortiz/chip8emulator/blob/985c6613190f33c351b23874a4091453c97048bf/chip8emulator/chip8emulator.py#L87C1-L93C26):
```py
    def on_draw(self):
        self.clear()
        for sprite in self.sprites:
            y = self.flip_and_scale_y_axis(sprite.center_y)
            x = self.scale_x_axis(sprite.center_x)
            pixel = self.graphics.get(x, y)
            sprite.color = arcade.color.WHITE if pixel == 1 else arcade.color.BLACK
            sprite.draw()
 ```
### 1. Separate the flip and color setting
These are state update operations.

### 2. Don't call draw on each sprite
Instead, call `self.sprites.draw()`.

**Comment by Cleptomania** [comment]
Oh yeah definitely only draw the SpriteList object directly. Iterating the sprites and calling draw on each one individually, is roughly the same cost performance wise for drawing each sprite as it is to draw all of them at once with the SpriteList.

**Comment by mmartinortiz** [comment]
> There seem to be some easy fixes you can make [here](https://github.com/mmartinortiz/chip8emulator/blob/985c6613190f33c351b23874a4091453c97048bf/chip8emulator/chip8emulator.py#L87C1-L93C26):
> 
> ```python
>     def on_draw(self):
>         self.clear()
>         for sprite in self.sprites:
>             y = self.flip_and_scale_y_axis(sprite.center_y)
>             x = self.scale_x_axis(sprite.center_x)
>             pixel = self.graphics.get(x, y)
>             sprite.color = arcade.color.WHITE if pixel == 1 else arcade.color.BLACK
>             sprite.draw()
> ```
> 
> ### 1. Separate the flip and color setting
> These are state update operations.
> 
> ### 2. Don't call draw on each sprite
> Instead, call `self.sprites.draw()`.

Wow, I missed that one completely, thank you very much. That really made things way better.

> I'd suggest, if you can, upgrading to Arcade 3.0 pre-releases, they're more stable than 2.6 anyways, and have a lot of performance improvements.

I'll take note of this and start using 3.0 :+1: 

Thank you both for the incredibly quick answers.

**Comment by einarf** [comment]
There is also a chip8 display shader example here working directly on a native "framebuffer" https://github.com/pythonarcade/arcade/blob/development/arcade/examples/gl/chip8_display.py

**Comment by mmartinortiz** [comment]
> There is also a chip8 display shader example here working directly on a native "framebuffer" https://github.com/pythonarcade/arcade/blob/development/arcade/examples/gl/chip8_display.py

Thanks. I've checked it out. I've always curious about frame buffers and this seems to be a perfect moment to try them.

Could you help me to understand why using a framebuffer directly in GL is better than using Arcade in my case?

**Comment by pushfoo** [comment]
**TL;DR:** Probably maximum speed.

The example einarf linked avoids many of the same things Arcade and pyglet's "hot" (frequently called) paths also avoid. Specifically, it has minimal overhead from dot access (`self.member.sub_member`) and function calls.

For games where you have multiple sprite lists and shaders, the overhead of those is worth it to avoid breaking GL state. For example, Arcade has context managers (things you can use in `with` blocks) to help set things back to the prior state so you don't break your rendering.

For an emulator, you might want to avoid that overhead. Many emulators avoid rendering the UI when not paused, and that may be relevant here too.

**Comment by einarf** [comment]
I don't know where framebuffers got into the picture here to be honest. The shader will stretch the content for you based on the configured projection. It could render the contents into a tiny framebuffer using chip8's physical "window size" and then stretch that texture across any screen size (adjusting for aspect ratio). For a pixelated effect we set the interpolation mode on the texture to `NEAREST` to get the desired effect.

Framebuffers are great in the following situations:
* Pixelated games with fixed resolution (render to fixed size framebuffer and stretch into the window). It means the player can use any window size or even go 4k and it will still look reasonable for that style.
* When you want to do post-processing of any kind. Color correction, lights (deferred), blur or anything else your imagination can come up with
* Other situations when a shader needs to access the pixel data for some other reason. It could be pixel-based collision testing, line of sight tracing, pathing or some form of masking.

**Comment by einarf** [comment]
The advantage of using the shader approach for the chip8 display:

I assume the emulator just modifies a ctypes/array.array/bytearray representing the framebuffer for the chip8 display. Every time this buffer changes (per frame) the data can be written to be GL buffer (`Buffer.write(data)`) to refresh the information on the GPU side. The shader rendering the chip8 framebuffer (in the linked example) will perform this job at a microsecond time scale leaving more room/time for for other code in your emulator.

Python is definitely not the fastest language and shaders can really help offloading work. The chip8 shader is also tested on GLES 3.2 (raspberry pi 4 and 5)

Also note that we have a Discord server. It's likely easier to share ideas there if you can access that platform.

**Comment by mmartinortiz** [comment]
I've been reading more about framebuffers and shaders (new topic for me, but that's why I got into a Chip-8, to learn new stuff) and, given your input and what I read, I understand that:

- The framebuffer is the memory area used to store a bitmap with the pixels to be drawn in the screen.
- Libraries like Arcade help to keep the whole GL state and, after processing all the game entities, will create a frame buffer.
- In the case of an emulator, the emulator itself and the software run by the emulator will do all the image processing: collision detection, shades, colors, etc.
- Because the emulator has its own framebuffer (where the emulated software puts the pixels for the screen) there is no need for Arcade to prepare the image again, and I can write directly into the `Buffer` class of Arcade.

However, I can still use Arcade's shader to take advantage of, for example, scaling the emulator's screen to the user's screen or get a CRT feel with the CRTFilter.

Did I get it right?

@einarf Thanks for the invitation to Discord. I prefer to keep the conversation here in GitHub, so it remains open to other people that may be interested in the future. Besides, when I tried Discord in the past, I never got used to it :-).

**Comment by einarf** [comment]
Actually I misspoke in the previous post. `Texture2D.write(..)` is the way to write the framebuffer data since it should go into a texture 😆. It could also go into a buffer but the texture API is simpler to work with in this case.

Just re-stating some things here to make sure we don't have any misunderstandings

I think that's a reasonable enough explanation. Arcade has a window with a framebuffer that can be rendered/drawn into. This is called the "default framebuffer". Offscreen framebuffers can also be created to draw into OpenGL textures (Texture2D) for further processing. Framebuffers are targets we can render into. They can have texture attachments meaning the contents of the framebuffer will end up in these textures. A single texture is enough in most use cases.

The chip8 example operates on framebuffer as represented in the memory of an 8bit microcomputer (32 * 64 bytes) and the shader converts with framebuffer format into the current RGBA pixel format. This framebuffer data is written into a integer texture and converted on the fly in the fragment shader. The reason for this is simply that the author of the emulator that sparked this example had an emulator working on the 32 * 64 framebuffer format. The pixel conversion was a bottleneck and we simply did that part in the shader.

If your emulator is using a 64 * 128 framebuffer (One byte per pixel appearing on the screen) you can make a 1-component opengl texture (`ctx.texture(..) creating a Texture2D`) and make a simple texture shader only doing the "stretching to the screen" part.

If I read everything right (except repeating my mistake about buffer vs texture writing) you are on the right path.

Possibly "framebuffer" is confusing here because we can mix up the chip8 framebuffer data in system memory and opengl framebuffer objects.

If you get stuck somewhere on shader part just ask.

**Comment by mmartinortiz** [comment]
I followed your advice and implemented a framebuffer to write directly the screen of the emulator. Now it is faster, but I still have to tune a few parts of my emulator (and fix some evident bugs).

So far, still enjoying this experience. Thanks for your help!
